# `pandas` Part 7: Combining Datasets with `concat()`

# Learning Objectives
## By the end of this tutorial you will be able to:
1. Combine DataFrames and/or Series with `concat()`
2. Understand a multi-index
3. Reset an index with `reset_index()`
4. Perform descriptive analytics on a combined DataFrame

## Files Needed for this lesson:
>- `CAvideos.csv`
>- `GBvideos.csv`
>- Download this csv files from Canvas prior to the lesson
>- C:\\Users\\mimc2537\\OneDrive - UCB-O365\\python\\pandas\\

## The general steps to working with pandas:
1. import pandas as pd
2. Create or load data into a pandas DataFrame or Series
3. Reading data with `pd.read_`
>- Excel files: `pd.read_excel('fileName.xlsx')`
>- Csv files: `pd.read_csv('fileName.csv')`
>- Note: if the file you want to read into your notebook is not in the same folder you can do one of two things:
>>- Move the file you want to read into the same folder/directory as the notebook
>>- Type out the full path into the read function
4. After steps 1-3 you will want to check out your DataFrame
>- Use `shape` to see how many records and columns are in your DataFrame
>- Use `head()` to show the first 5-10 records in your DataFrame

# Introduction Notes on Combining Data Using `pandas`
1. Being able to combine data from multiple sources is a critical skill for analytics professionals
2. We will learn the `pandas` way of combining data but there are similarities here to SQL
3. Why combine data with `pandas` if you can do the same thing in SQL?
>- The answer to this depends on the project
>- Some projects may be completed more efficiently all with `pandas` so you wouldn't necessarily need SQL
>- For some projects incorporating SQL into our python code makes sense
>- In a an analytics job, you will likely use both python and SQL to get the job done! 

# Initial set-up steps
1. import modules and check working directory
2. Read data in
3. Check the data

In [1]:
import os
import pandas as pd

os.getcwd()
for file in os.listdir():
    if '.csv' in file:
        print(file)
    elif '.xlsx' in file:
        print(file)
        
file1 = "CAvideos.csv"
file2 = "GBvideos.csv"


CAvideos.csv
GBvideos.csv
winemag-data-130k-v2 (2).csv


# Step 2 Read Data Into a DataFrame with `read_csv()`
>- file names: 
>>- `CAvideos.csv`
>>- `GBvideos.csv`

In [2]:
ca = pd.read_csv(file1)
gb = pd.read_csv(file2)

### Check how many rows and columns are in our DataFrames

In [3]:
ca.head(3)

video_id trending_date  \
0  n1WpP7iowLc      17.14.11   
1  0dBIkQ4Mz1M      17.14.11   
2  5qpjK5DgCt4      17.14.11   

                                               title channel_title  \
0         Eminem - Walk On Water (Audio) ft. Beyoncé    EminemVEVO   
1                      PLUSH - Bad Unboxing Fan Mail     iDubbbzTV   
2  Racist Superman | Rudy Mancuso, King Bach & Le...  Rudy Mancuso   

   category_id              publish_time  \
0           10  2017-11-10T17:00:03.000Z   
1           23  2017-11-13T17:00:00.000Z   
2           23  2017-11-12T19:05:24.000Z   

                                                tags     views   likes  \
0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579  787425   
1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651  127794   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434  146035   

   dislikes  comment_count                                  thumbnail_link  \
0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   

                                         description  
0  Eminem's new track Walk on Water ft. Beyoncé i...  
1  STill got a lot of packages. Probably will las...  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...

In [4]:
gb.head(3)

video_id trending_date                                          title  \
0  Jw1Y-zhQURU      17.14.11  John Lewis Christmas Ad 2017 - #MozTheMonster   
1  3s1rvMFUweQ      17.14.11      Taylor Swift: …Ready for It? (Live) - SNL   
2  n1WpP7iowLc      17.14.11     Eminem - Walk On Water (Audio) ft. Beyoncé   

         channel_title  category_id              publish_time  \
0           John Lewis           26  2017-11-10T07:38:29.000Z   
1  Saturday Night Live           24  2017-11-12T06:24:44.000Z   
2           EminemVEVO           10  2017-11-10T17:00:03.000Z   

                                                tags     views   likes  \
0  christmas|"john lewis christmas"|"john lewis"|...   7224515   55681   
1  SNL|"Saturday Night Live"|"SNL Season 43"|"Epi...   1053632   25561   
2  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579  787420   

   dislikes  comment_count                                  thumbnail_link  \
0     10247           9479  https://i.ytimg.com/vi/Jw1Y-zhQURU/default.jpg   
1      2294           2757  https://i.ytimg.com/vi/3s1rvMFUweQ/default.jpg   
2     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   

                                         description  
0  Click here to continue the story and make your...  
1  Musical guest Taylor Swift performs …Ready for...  
2  Eminem's new track Walk on Water ft. Beyoncé i...

### Check a couple of rows of data in one of the new DataFrames

In [5]:
print(ca.shape, gb.shape)

(40881, 16) (38916, 16)


## Check the datatypes

In [6]:
gb.dtypes

video_id                  object
trending_date             object
title                     object
channel_title             object
category_id                int64
publish_time              object
tags                      object
views                      int64
likes                      int64
dislikes                   int64
comment_count              int64
thumbnail_link            object
comments_disabled           bool
ratings_disabled            bool
video_error_or_removed      bool
description               object
dtype: object

# Combining DataFrames
>- The three common ways to combine datastest in pandas is with `concat()`, `join()`, and `merge()`
>- `concat()` will take two DataFrames or Series and append them together
>>- This is basically taking DataFrames and stacking their data on top of each other into one DataFrame
>>- For `concat()` you need the columns/fields in both DataFrames to the be the same
>- `join()` "links" DataFrames together based on a common field/column between the two
>- `merge()` also links DataFrames together based on common field/columns but with different syntax.
>>- We will cover the most basic join in this class
>>- A more in depth study of joins is provided in SQL focused courses
>>- Pandas join reference for further study: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html


# Using the YouTube DataFrames to practice combining data with pandas
>- The YouTube datasets store data on various YouTube trending statistics
>- Our example datasets show several months of data and daily trending YouTube videos.

>- For more information and for other YouTube datasets see the following link:
>>- https://www.kaggle.com/datasnaek/youtube-new

### First, creating a new DataFrame that appends the Canadian and British YouTube DataFrames

In [7]:
CanUK = pd.concat([ca, gb], keys=['CA','GB'])
print(CanUK.shape, ca.shape, gb.shape)

(79797, 16) (40881, 16) (38916, 16)


#### Some notes on the previous code
>- Line 1: We define a new DataFrame named `CanUK` which is defined as the concatenation,`concat()`, of two datasets
>>- Dataset 1 = canadian_youtube
>>- Dataset 2 = uk_youtube
>>- The `concat()` function takes the two (or more if applicable) DataFrames and "stacks" them on top of each
>- Line 2: We use `keys` option to define a multi-index (aka hierarchical index)
>>- Because our datasets represent YouTube videos from different countries we pass the abbreviated names of those countries as a list to `keys`
>>- Enter the keys names in order they appear in line 1 (e.g., 'can' first, 'uk' second)
>- Line 3: We use the `names` option to label our index columns from line 2
>>- Without the `names` option we would not have anything above our index columns

### Check the index for any dataframe using `DataFrame.index`
>- Note how `concat()` uses the rowid's for each country's dataset versus continuing the count

In [8]:
CanUK.index

MultiIndex([('CA',     0),
            ('CA',     1),
            ('CA',     2),
            ('CA',     3),
            ('CA',     4),
            ('CA',     5),
            ('CA',     6),
            ('CA',     7),
            ('CA',     8),
            ('CA',     9),
            ...
            ('GB', 38906),
            ('GB', 38907),
            ('GB', 38908),
            ('GB', 38909),
            ('GB', 38910),
            ('GB', 38911),
            ('GB', 38912),
            ('GB', 38913),
            ('GB', 38914),
            ('GB', 38915)],
           length=79797)

### Take a look at our new DataFrame

In [9]:
CanUK.head(3)

video_id trending_date  \
CA 0  n1WpP7iowLc      17.14.11   
   1  0dBIkQ4Mz1M      17.14.11   
   2  5qpjK5DgCt4      17.14.11   

                                                  title channel_title  \
CA 0         Eminem - Walk On Water (Audio) ft. Beyoncé    EminemVEVO   
   1                      PLUSH - Bad Unboxing Fan Mail     iDubbbzTV   
   2  Racist Superman | Rudy Mancuso, King Bach & Le...  Rudy Mancuso   

      category_id              publish_time  \
CA 0           10  2017-11-10T17:00:03.000Z   
   1           23  2017-11-13T17:00:00.000Z   
   2           23  2017-11-12T19:05:24.000Z   

                                                   tags     views   likes  \
CA 0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579  787425   
   1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651  127794   
   2  racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434  146035   

      dislikes  comment_count                                  thumbnail_link  \
CA 0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
   1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   
   2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   

      comments_disabled  ratings_disabled  video_error_or_removed  \
CA 0              False             False                   False   
   1              False             False                   False   
   2              False             False                   False   

                                            description  
CA 0  Eminem's new track Walk on Water ft. Beyoncé i...  
   1  STill got a lot of packages. Probably will las...  
   2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...

#### Did using `concat()` work to append the two DataFrames together? 
>- Check the shape of your new DataFrame
>- Compare the number of records to each one individually
>>- canadian_youtube = 40881 records
>>- uk_youtube = 38916 records
>>- 40881 + 38916 = 79797 total records

#### `reset_index`:
##### Note: You can reset a an index with `reset_index` 
>- This can be useful for some situations
>- For a multi-index you can pass the `level` option and specify what index you want to reset
>- Note: To make the change to our current DataFrame we would need to use the option, `inplace=True`

In [10]:
CanUK.reset_index(level = 0, inplace = True)

CanUK.head(3)

level_0     video_id trending_date  \
0      CA  n1WpP7iowLc      17.14.11   
1      CA  0dBIkQ4Mz1M      17.14.11   
2      CA  5qpjK5DgCt4      17.14.11   

                                               title channel_title  \
0         Eminem - Walk On Water (Audio) ft. Beyoncé    EminemVEVO   
1                      PLUSH - Bad Unboxing Fan Mail     iDubbbzTV   
2  Racist Superman | Rudy Mancuso, King Bach & Le...  Rudy Mancuso   

   category_id              publish_time  \
0           10  2017-11-10T17:00:03.000Z   
1           23  2017-11-13T17:00:00.000Z   
2           23  2017-11-12T19:05:24.000Z   

                                                tags     views   likes  \
0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579  787425   
1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651  127794   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434  146035   

   dislikes  comment_count                                  thumbnail_link  \
0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   

                                         description  
0  Eminem's new track Walk on Water ft. Beyoncé i...  
1  STill got a lot of packages. Probably will las...  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...

In [11]:
CanUK = CanUK.rename(columns = {'level_0':'Country'})
CanUK

Country     video_id trending_date  \
0          CA  n1WpP7iowLc      17.14.11   
1          CA  0dBIkQ4Mz1M      17.14.11   
2          CA  5qpjK5DgCt4      17.14.11   
3          CA  d380meD0W0M      17.14.11   
4          CA  2Vv-BfVoq4g      17.14.11   
...       ...          ...           ...   
38911      GB  l884wKofd54      18.14.06   
38912      GB  IP8k2xkhOdI      18.14.06   
38913      GB  Il-an3K9pjg      18.14.06   
38914      GB  -DRsfNObKIQ      18.14.06   
38915      GB  4YFo4bdMO8Q      18.14.06   

                                                   title  \
0             Eminem - Walk On Water (Audio) ft. Beyoncé   
1                          PLUSH - Bad Unboxing Fan Mail   
2      Racist Superman | Rudy Mancuso, King Bach & Le...   
3                               I Dare You: GOING BALD!?   
4            Ed Sheeran - Perfect (Official Music Video)   
...                                                  ...   
38911  Enrique Iglesias - MOVE TO MIAMI (Official Vid...   
38912  Jacob Sartorius - Up With It (Official Music V...   
38913                 Anne-Marie - 2002 [Official Video]   
38914  Eleni Foureira - Fuego - Cyprus - LIVE - First...   
38915  KYLE - Ikuyo feat.  2 Chainz & Sophia Black [A...   

                 channel_title  category_id              publish_time  \
0                   EminemVEVO           10  2017-11-10T17:00:03.000Z   
1                    iDubbbzTV           23  2017-11-13T17:00:00.000Z   
2                 Rudy Mancuso           23  2017-11-12T19:05:24.000Z   
3                     nigahiga           24  2017-11-12T18:01:41.000Z   
4                   Ed Sheeran           10  2017-11-09T11:04:14.000Z   
...                        ...          ...                       ...   
38911      EnriqueIglesiasVEVO           10  2018-05-09T07:00:01.000Z   
38912          Jacob Sartorius           10  2018-05-11T17:09:16.000Z   
38913               Anne-Marie           10  2018-05-08T11:05:08.000Z   
38914  Eurovision Song Contest           24  2018-05-08T20:32:32.000Z   
38915           SuperDuperKyle           10  2018-05-11T04:06:35.000Z   

                                                    tags     views    likes  \
0      Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579   787425   
1      plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651   127794   
2      racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434   146035   
3      ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   2095828   132239   
4      edsheeran|"ed sheeran"|"acoustic"|"live"|"cove...  33523622  1634130   
...                                                  ...       ...      ...   
38911  Enrique Iglesias feat. Pitbull|"MOVE TO MIAMI"...  25066952   268088   
38912  jacob sartorius|"jacob"|"up with it"|"jacob sa...   1492219    61998   
38913  anne|"marie"|"anne-marie"|"2002"|"two thousand...  29641412   394830   
38914  Eurovision Song Contest|"2018"|"Lisbon"|"Cypru...  14317515   151870   
38915  Kyle|"SuperDuperKyle"|"Ikuyo"|"2 Chainz"|"Soph...    607552    18271   

       dislikes  comment_count  \
0         43420         125882   
1          1688          13030   
2          5339           8181   
3          1989          17518   
4         21082          85067   
...         ...            ...   
38911     12783           9933   
38912     13781          24330   
38913      8892          19988   
38914     45875          26766   
38915       274           1423   

                                       thumbnail_link  comments_disabled  \
0      https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg              False   
1      https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg              False   
2      https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg              False   
3      https://i.ytimg.com/vi/d380meD0W0M/default.jpg              False   
4      https://i.ytimg.com/vi/2Vv-BfVoq4g/default.jpg              False   
...                                               ...            

# Now some descriptive analytics

### What channels have the most trending videos?

In [13]:
CanUK.groupby('channel_title').channel_title.count().sort_values(ascending = False)

channel_title
The Late Show with Stephen Colbert        361
TheEllenShow                              357
Jimmy Kimmel Live                         351
The Tonight Show Starring Jimmy Fallon    345
Breakfast Club Power 105.1 FM             342
                                         ... 
Ghost Adventures Season 16                  1
Global Savage                               1
GloomyHouse                                 1
GoCanucksGo                                 1
활력소TV                                       1
Name: channel_title, Length: 5956, dtype: int64

### What are the quantitative descriptive statistics for TheEllenShow?

In [14]:
CanUK.groupby('channel_title').views.sum().sort_values(ascending = False)

channel_title
NickyJamTV                    8758841250
Ozuna                         8447128427
Bad Bunny                     6981462307
DrakeVEVO                     6772018167
ChildishGambinoVEVO           6513084682
                                 ...    
Boston Celtics on MassLive          1637
Qc TV HD                            1569
Mathieu Désy                        1464
Georgia Webster                     1187
udearroba                           1141
Name: views, Length: 5956, dtype: int64

In [17]:
#ellen show answer

CanUK[CanUK['channel_title'] == 'TheEllenShow'].describe()

category_id         views          likes      dislikes  comment_count
count        357.0  3.570000e+02     357.000000    357.000000     357.000000
mean          24.0  1.779820e+06   46257.770308   1328.537815    2179.221289
std            0.0  2.008212e+06   62550.222512   3198.710961    3614.375611
min           24.0  5.631300e+04    1187.000000     18.000000       0.000000
25%           24.0  5.360520e+05   14152.000000    243.000000     318.000000
50%           24.0  1.238747e+06   28108.000000    598.000000     940.000000
75%           24.0  2.022094e+06   59950.000000   1309.000000    2858.000000
max           24.0  1.359265e+07  499673.000000  26397.000000   28381.000000

##### Alternatively, you can use `loc[]` to peform the filtering operation
>- The use of `where()` or `loc[]` depends on the question/purpose or sometimes just personal preference

In [18]:
CanUK[CanUK['channel_title'] == 'TheEllenShow'].describe()

category_id         views          likes      dislikes  comment_count
count        357.0  3.570000e+02     357.000000    357.000000     357.000000
mean          24.0  1.779820e+06   46257.770308   1328.537815    2179.221289
std            0.0  2.008212e+06   62550.222512   3198.710961    3614.375611
min           24.0  5.631300e+04    1187.000000     18.000000       0.000000
25%           24.0  5.360520e+05   14152.000000    243.000000     318.000000
50%           24.0  1.238747e+06   28108.000000    598.000000     940.000000
75%           24.0  2.022094e+06   59950.000000   1309.000000    2858.000000
max           24.0  1.359265e+07  499673.000000  26397.000000   28381.000000

In [20]:
CanUK.where(CanUK['channel_title'] == 'TheEllenShow').describe()

category_id         views          likes      dislikes  comment_count  \
count        357.0  3.570000e+02     357.000000    357.000000     357.000000   
mean          24.0  1.779820e+06   46257.770308   1328.537815    2179.221289   
std            0.0  2.008212e+06   62550.222512   3198.710961    3614.375611   
min           24.0  5.631300e+04    1187.000000     18.000000       0.000000   
25%           24.0  5.360520e+05   14152.000000    243.000000     318.000000   
50%           24.0  1.238747e+06   28108.000000    598.000000     940.000000   
75%           24.0  2.022094e+06   59950.000000   1309.000000    2858.000000   
max           24.0  1.359265e+07  499673.000000  26397.000000   28381.000000   

       comments_disabled  ratings_disabled  video_error_or_removed  
count         357.000000             357.0                   357.0  
mean            0.170868               0.0                     0.0  
std             0.376922               0.0                     0.0  
min             0.000000               0.0                     0.0  
25%             0.000000               0.0                     0.0  
50%             0.000000               0.0                     0.0  
75%             0.000000               0.0                     0.0  
max             1.000000               0.0                     0.0

### What were the total YouTube videos, total views, likes and dislikes for TheEllenShow?
>- Using the agg() function to calculate specific aggregations on different columns

In [24]:
CanUK[CanUK['channel_title'] == 'TheEllenShow']\
.agg({'video_id':['count'],'views':['sum'],'likes':['sum','mean'],'dislikes':['sum']})

video_id        views         likes  dislikes
count     357.0          NaN           NaN       NaN
sum         NaN  635395636.0  1.651402e+07  474288.0
mean        NaN          NaN  4.625777e+04       NaN

## What are the totals for TheEllenShow's top 5 most viewed videos?
>- Only include the title names as part of the output (not channel or any other categorical fields)
>- Include total views, likes, dislikes, and comment count in the output

In [ ]:
CanUK
#dat.col.sum() basically

In [29]:
CanUK[CanUK.channel_title == 'TheEllenShow']\
.groupby('title')\
[['title','views','likes','dislikes','comment_count']]\
.sum()\
.sort_values(by = 'views',ascending = False)[0:5]

views    likes  \
title                                                                   
Jennifer Lawrence Explains Her Drunk Alter Ego ...  71621685  1302161   
Kid Yodeler Mason Ramsey Performs                   68553312  2697959   
Kim Kardashian Lets Gender of Third Child Slip      52985123   776177   
Billionaire Bill Gates Guesses Grocery Store Pr...  37076353   542804   
Laurel or Yanny?                                    28695358   613487   

                                                    dislikes  comment_count  
title                                                                        
Jennifer Lawrence Explains Her Drunk Alter Ego ...     35255          49762  
Kid Yodeler Mason Ramsey Performs                     144604              0  
Kim Kardashian Lets Gender of Third Child Slip         26509          50940  
Billionaire Bill Gates Guesses Grocery Store Pr...     29036          42720  
Laurel or Yanny?                                       18568         100731

# Some Notes on the Previous Example
>- Our pandas code in the previous example is similar to SQL in the following ways
    1. `loc[CanUk.channel_title == 'TheEllenShow',` is SQL equivalent to `WHERE channel_title = 'TheEllenShow'`
    2. `['title','views','likes','dislikes','comment_count']` is SQL equivalent to:
        `SELECT title, sum(views),sum(likes),sum(dislikes),sum(comment_count)`
    3. `groupby(['title`]) is SQL equivalent to GROUP BY title
    4. Now in pandas we enter the aggregation after the `groupby()`, in this example `sum()`
      >>- In SQL we write the aggregation in the SELECT statement
      
## In future lessons we will continue to learn how pandas and SQL relate